### PR TITLE
RmlUI Context handling improvements

### LINKED ISSUE #1832: EngineCrash after reloading rmlui widget
<html><body>

module | address | file | line
-- | -- | -- | --
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x140cabcc6 | /mxe/usr/lib/gcc/x86_64-w64-mingw32.static/13.2.0/include/c++/bits/stl_iterator.h | 1077
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x140c8a8b3 | /mxe/usr/lib/gcc/x86_64-w64-mingw32.static/13.2.0/include/c++/bits/vector.tcc | 114
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x140f50e1b | rts/lib/sol2/sol.hpp | 16191
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x140fe182c | rts/lib/sol2/sol.hpp | 8398
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x1408f9c0d | rts/lib/lua/src/ldo.cpp | 321
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x14090b893 | rts/lib/lua/src/lvm.cpp | 620
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x1408fa015 | rts/lib/lua/src/ldo.cpp | 378
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x1408fa3b9 | rts/lib/lua/src/ldo.cpp | 119
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x1408f2c57 | rts/lib/lua/src/lapi.cpp | 833
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x1408f573e | rts/lib/lua/src/lbaselib.cpp | 390
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x1408f9c0d | rts/lib/lua/src/ldo.cpp | 321
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x14090b893 | rts/lib/lua/src/lvm.cpp | 620
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x1408fa015 | rts/lib/lua/src/ldo.cpp | 378
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x1408fa3b9 | rts/lib/lua/src/ldo.cpp | 119
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x1408f2c57 | rts/lib/lua/src/lapi.cpp | 833
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x14021dfaa | rts/Lua/LuaHandle.cpp | 363
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x140222c4a | rts/Lua/LuaHandle.cpp | 458
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x14022ccdb | rts/Lua/LuaHandle.h | 394
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x140709b19 | rts/System/EventHandler.cpp | 828
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x140135c27 | rts/Game/UI/MouseHandler.cpp | 513
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x14071f445 | rts/System/Input/MouseInput.cpp | 86
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x14071d870 | rts/System/Input/InputHandler.cpp | 14
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x14075b795 | /mxe/usr/lib/gcc/x86_64-w64-mingw32.static/13.2.0/include/c++/bits/basic_string.h | 223
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x140739c0b | rts/System/Main.cpp | 50
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x1400012ee | /mxe/tmp-gcc-x86_64-w64-mingw32.static/gcc-13.2.0.build_/mingw-w64-v11.0.0/mingw-w64-crt/crt/crtexe.c | 273
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe | 0x1400013e6 | /mxe/tmp-gcc-x86_64-w64-mingw32.static/gcc-13.2.0.build_/mingw-w64-v11.0.0/mingw-w64-crt/crt/crtexe.c | 164
C:\Windows\SYSTEM32\ntdll.dll | 0x000000000004cc91 | ?? | 0

</body>
</html>module	address	file	line
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x140cabcc6	/mxe/usr/lib/gcc/x86_64-w64-mingw32.static/13.2.0/include/c++/bits/stl_iterator.h 	1077
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x140c8a8b3	/mxe/usr/lib/gcc/x86_64-w64-mingw32.static/13.2.0/include/c++/bits/vector.tcc 	114
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x140f50e1b	rts/lib/sol2/sol.hpp 	[16191](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/sol2/sol.hpp#L16191)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x140fe182c	rts/lib/sol2/sol.hpp 	[8398](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/sol2/sol.hpp#L8398)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x1408f9c0d	rts/lib/lua/src/ldo.cpp 	[321](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/lua/src/ldo.cpp#L321)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x14090b893	rts/lib/lua/src/lvm.cpp 	[620](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/lua/src/lvm.cpp#L620)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x1408fa015	rts/lib/lua/src/ldo.cpp 	[378](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/lua/src/ldo.cpp#L378)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x1408fa3b9	rts/lib/lua/src/ldo.cpp 	[119](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/lua/src/ldo.cpp#L119)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x1408f2c57	rts/lib/lua/src/lapi.cpp 	[833](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/lua/src/lapi.cpp#L833)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x1408f573e	rts/lib/lua/src/lbaselib.cpp 	[390](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/lua/src/lbaselib.cpp#L390)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x1408f9c0d	rts/lib/lua/src/ldo.cpp 	[321](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/lua/src/ldo.cpp#L321)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x14090b893	rts/lib/lua/src/lvm.cpp 	[620](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/lua/src/lvm.cpp#L620)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x1408fa015	rts/lib/lua/src/ldo.cpp 	[378](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/lua/src/ldo.cpp#L378)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x1408fa3b9	rts/lib/lua/src/ldo.cpp 	[119](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/lua/src/ldo.cpp#L119)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x1408f2c57	rts/lib/lua/src/lapi.cpp 	[833](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/lib/lua/src/lapi.cpp#L833)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x14021dfaa	rts/Lua/LuaHandle.cpp 	[363](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/Lua/LuaHandle.cpp#L363)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x140222c4a	rts/Lua/LuaHandle.cpp 	[458](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/Lua/LuaHandle.cpp#L458)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x14022ccdb	rts/Lua/LuaHandle.h 	[394](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/Lua/LuaHandle.h#L394)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x140709b19	rts/System/EventHandler.cpp 	[828](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/System/EventHandler.cpp#L828)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x140135c27	rts/Game/UI/MouseHandler.cpp 	[513](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/Game/UI/MouseHandler.cpp#L513)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x14071f445	rts/System/Input/MouseInput.cpp 	[86](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/System/Input/MouseInput.cpp#L86)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x14071d870	rts/System/Input/InputHandler.cpp 	[14](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/System/Input/InputHandler.cpp#L14)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x14075b795	/mxe/usr/lib/gcc/x86_64-w64-mingw32.static/13.2.0/include/c++/bits/basic_string.h 	223
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x140739c0b	rts/System/Main.cpp 	[50](http://github.com/beyond-all-reason/spring/tree/BAR105/rts/System/Main.cpp#L50)
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x1400012ee	/mxe/tmp-gcc-x86_64-w64-mingw32.static/gcc-13.2.0.build_/mingw-w64-v11.0.0/mingw-w64-crt/crt/crtexe.c 	273
C:\Users\user\AppData\Local\Programs\Beyond-All-Reason\data\engine\105.1.1-2708-g3d35645 bar\spring.exe	0x1400013e6	/mxe/tmp-gcc-x86_64-w64-mingw32.static/gcc-13.2.0.build_/mingw-w64-v11.0.0/mingw-w64-crt/crt/crtexe.c 	164
C:\Windows\SYSTEM32\ntdll.dll	0x000000000004cc91	?? 	0

I only changed the following code in the example:
```html
		<label>
			<h4>Range Input</h4>
			<input type="range" min="0" max="100" step="1"/>
			<h4>Range Input</h4>

			<input type="range" min="0" max="100" step="1"/>
			<input type="range" min="0" max="100" step="1"/>
		</label>

```

### PR DESCRIPTION
**Pull request by ChrisFloofyKitsune**
context handling improvements

* manage contexts in the "contexts_to_remove" hashset more carefully to prevent use-after-free crash.
* Remove context from removal hashset if it is "recreated" in the same frame.

use ranges::rotate to reorder clicked context

fixes #1832

### CODE REVIEW

File: rts/Rml/Backends/RmlUi_Backend.cpp [hunk 1]
```diff
@@ -308,6 +308,34 @@ void RmlGui::OnContextDestroy(Rml::Context* context)
...
+Rml::Context* RmlGui::GetOrCreateContext(const std::string& name)
+{
+	if (!RmlInitialized()) {
+		return nullptr;
+	}
+	
+	Rml::Context* context = Rml::GetContext(name);
+	if (context == nullptr) {
+		context = Rml::CreateContext(name, {0, 0});
+	} else if (state->contexts_to_remove.contains(context)) {
```
Comment (sprunk) [inline rts/Rml/Backends/RmlUi_Backend.cpp:320]:
```suggestion
	} else if (state->contexts_to_remove.contains(context)) {
		// can happen if name reused on the same frame
```

File: rts/Rml/Backends/RmlUi_Backend.cpp [hunk 2]
```diff
@@ -308,6 +308,34 @@ void RmlGui::OnContextDestroy(Rml::Context* context)
...
+		state->contexts_to_remove.erase(context);
```
Comment (sprunk) [inline rts/Rml/Backends/RmlUi_Backend.cpp:321]: Maybe we could do without the `.contains` check? `.erase` has to search for the element anyway so we'd avoid searching it twice if present and won't see any difference if not.

Comment (ChrisFloofyKitsune) [reply rts/Rml/Backends/RmlUi_Backend.cpp:321]: I didn't realize I could do that

**Comment by lhog** [comment]
@ChrisFloofyKitsune do you plan any changes (as perhaps per @sprunk comments above)?

**Comment by Beherith** [comment]
Indeed, I'm quite eager to get my hands dirty with rmlui over the xmas break <3

**Review by sprunk — APPROVED** [review]

**Review by sprunk — APPROVED** [review]

**Comment by Beherith** [comment]
Tested and fixes the crash. Yay!

**Pull request merged**